sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+        # values (equal to (n_clusters * n_features)) comparatively to the global
+        # number of work items, i.e. when the probability of collision is high. At the
+        # opposite end where the probability of collision is low, privatization might
+        # be detrimental to performance and we might prefer simpler, faster code
+        # with updates directly made into the final array of centroids.
+
+        # The privatization strategy also applies to the updates of the centroid
+        # counts.
+
+        if sample_idx >= n_samples:
+            return
+
+        per_sample_pseudo_inertia[sample_idx] = min_pseudo_inertia
+
+        # each work item is assigned an array of centroids in a round robin manner
+        privatization_idx = (
+            sample_idx // preferred_work_group_size_multiple
+        ) % n_centroids_private_copies
+
+        dpex.atomic.add(
+            cluster_sizes_private_copies,
+            (privatization_idx, min_idx),
+            one,
+        )
+
+        for feature_idx in range(n_features):
+            dpex.atomic.add(
+                new_centroids_t_private_copies,
+                (privatization_idx, feature_idx, min_idx),
+                X_t[feature_idx, sample_idx],
+            )
+
+    return (
+        n_centroids_private_copies,
+        fused_kmeans[global_size, work_group_size],
+    )
+
+
+@lru_cache
+def make_assignment_fixed_window_kernel(
+    n_samples,
+    n_features,
+    n_clusters,
+    preferred_work_group_size_multiple,
+    centroids_window_width_multiplier,
+    centroids_window_height,
+    work_group_size,
+    dtype,
+):
+    window_n_centroids = (
+        preferred_work_group_size_multiple * centroids_window_width_multiplier
+    )
+    n_window_features_per_work_group = work_group_size // window_n_centroids
+    centroids_window_height_ratio_multiplier = (
+        centroids_window_height // n_window_features_per_work_group
+    )
+
+    window_n_features = centroids_window_height
+
+    n_windows_per_feature = math.ceil(n_clusters / window_n_centroids)
+    global_size = (math.ceil(n_samples / work_group_size)) * (work_group_size)
+    n_windows_per_centroid = math.ceil(n_features / window_n_features)
+    centroids_window_shape = (window_n_features, (window_n_centroids + 1))
+
+    inf = dtype(math.inf)
+    zero = dtype(0.0)
+    two = dtype(2)
+
+    # TODO: factorize the fused fixed kernel and the assignment kernel using dpex.func
+    # namespace. Ensure no impact on performance.
+    # fmt: off
+    @dpex.kernel
+    def assignment(
+        X_t,                                # IN    (n_features, n_samples)
+        current_centroids_t,                # IN    (n_features, n_clusters)
+        centroids_half_l2_norm,             # IN    (n_clusters,)
+        per_sample_inertia,                 # OUT   (n_samples,)
+        assignments_idx,                    # OUT   (n_samples,)
```

Comment:
Thanks for indicating shapes in comment.

Small suggestion: If tensors are only meant to be read, this can be adapt as part of the comment as well, [similarly to the indication in scikit-learn's implementations](https://github.com/scikit-learn/scikit-learn/blob/7bdd426704fc49316b3273fb33de7f6931c22692/sklearn/cluster/_k_means_lloyd.pyx#L24-L26) for instance.